File: 백준/Gold/14719. 빗물/빗물.py
import sys
input = sys.stdin.readline

H, W = map(int, input().split())
lst = [0] + list(map(int, input().split())) + [0] # 양 끝에 0더해주기

# 왼쪽 최대값 저장할 배열
prefix_max = [0] * (W+2)
# 오른쪽 최대값 저장할 배열
suffix_max = [0] * (W+2)

# 왼쪽부터 돌면서 최대값을 저장하기
for i in range(1, W+1):
    prefix_max[i] = max(prefix_max[i-1], lst[i])

# 오른쪽부터 돌면서 최대값을 저장하기
for i in range(W, 0, -1):
    suffix_max[i] = max(suffix_max[i+1], lst[i])

answer = 0
# 왼쪽부터 돌면서
for i in range(W):
    # 물이고이는양 = (왼쪽, 오른쪽 최대값 중 최솟값 - 원래의 블록) 을 answer에 더해주기 
   answer += min(prefix_max[i], suffix_max[i]) - lst[i]

print(answer)

File: 백준/Gold/14719. 빗물/빗물.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/**
창고다각형과 비슷한 문제

어떤 위치에서 빗물의 높이는 왼쪽에서 최고 기둥과 오른쪽에서 최고 기둥 중 가장 낮은 기둥의 높이이다

대신 현재 기둥의 높이는 빼기
 */
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		st = new StringTokenizer(br.readLine());
		int H = Integer.parseInt(st.nextToken());
		int W = Integer.parseInt(st.nextToken());
		
		int[] arr = new int[W+1];
		// 왼쪽에서 최고 기둥
		int[] prefix_max = new int[W+2];
		// 오른쪽에서 최고 기둥
		int[] suffix_max = new int[W+2];
		
		st = new StringTokenizer(br.readLine());
		for (int i = 1; i < W+1; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		} // 기둥 입력 받기
		
		// 왼쪽에서 최고 기둥 구하기
		for (int i = 1; i < W+1; i++) {
			prefix_max[i] = Math.max(prefix_max[i-1], arr[i]);
		}
		
		// 오른쪽에서 최고 기둥 구하기
		for (int i = W; i > 0; i--) {
			suffix_max[i] = Math.max(suffix_max[i+1], arr[i]);
		}
		
		int answer = 0;
		// 왼쪽에서 최고 기둥과 오른쪽에서 최고 기둥 중 가장 낮은 기둥의 높이 에서 현재의 기둥 높이 빼기
		for (int i = 1; i < W; i++) {
			answer += Math.min(prefix_max[i], suffix_max[i]) - arr[i];
		}
		
		System.out.println(answer);
		
	} // end of main
} // end of class